Add an extension to DrakeLcmInterface to support log file generation and playback.

siyuanfeng-tri opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/7178)
<!-- Reviewable:end -->

siyuanfeng-tri commented:
+@naveenoid for feature review. Thanks!
cc @ericcousineau-tri @russtedrake 

---

Review status: 0 of 14 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-:-KvbhXeV1dPLKALYKhvm:b-wyb0z4)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 0 of 14 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/lcm/drake_lcm_interface.h, line 56 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kvbty_sRPeSguEhdpS6:-Kvbty_sRPeSguEhdpS7:b-k7ykok) ([raw file](https://github.com/robotlocomotion/drake/blob/215bd2c3052209bec95c49a292eafbb597f32d5a/drake/lcm/drake_lcm_interface.h#L56)):*
> ```objc
>    */
>   virtual void Publish(const std::string& channel, const void* data,
>                        int data_size, double time_sec = 0) = 0;
> ```

Rather than having a per-call publish specification for time, could there be something akin to ROS's `/clock` when `/use_sim_time` is true?

e.g. Clock everything off of a single message, published at every step of the simulation.

---

*[drake/lcm/drake_lcm_log.h, line 38 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-KvbtfgmxYEbbjQ_ZaRC:-KvbtfgmxYEbbjQ_ZaRD:b-5i6sj0) ([raw file](https://github.com/robotlocomotion/drake/blob/215bd2c3052209bec95c49a292eafbb597f32d5a/drake/lcm/drake_lcm_log.h#L38)):*
> ```objc
>    * @throws std::runtime_error if unable to open file.
>    */
>   explicit DrakeLcmLog(const std::string& file_name, bool is_write);
> ```

Is there a way to add a connection to publish to another `DrakeLcmInterface` (namely `DrakeLcm`) such that it is easier to inspect the behavior of the playback?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7178)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 0 of 14 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/lcm/drake_lcm_interface.h, line 56 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kvbty_sRPeSguEhdpS6:-KvbuKoh2t3qX28-550S:bodzl4q) ([raw file](https://github.com/robotlocomotion/drake/blob/215bd2c3052209bec95c49a292eafbb597f32d5a/drake/lcm/drake_lcm_interface.h#L56)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Rather than having a per-call publish specification for time, could there be something akin to ROS's `/clock` when `/use_sim_time` is true?

e.g. Clock everything off of a single message, published at every step of the simulation.
</blockquote></details>

As an alternative, it'd be nice if a similar architecture could be used for running a simulation against a `rosbag` - from what I understand, playback could probably be managed such that it would use the timestamps made according to system time, or according to publishes on `/clock`.

http://wiki.ros.org/Clock

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7178)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
*[drake/lcm/drake_lcm_interface.h, line 56 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kvbty_sRPeSguEhdpS6:-Kvbupkbk9hn_U7zrdHm:bld0pa5) ([raw file](https://github.com/robotlocomotion/drake/blob/215bd2c3052209bec95c49a292eafbb597f32d5a/drake/lcm/drake_lcm_interface.h#L56)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

As an alternative, it'd be nice if a similar architecture could be used for running a simulation against a `rosbag` - from what I understand, playback could probably be managed such that it would use the timestamps made according to system time, or according to publishes on `/clock`.

http://wiki.ros.org/Clock
</blockquote></details>

I'd want to keep it simple.. at first, also I don't think the system framework is designed for distributed whatever, so there is really no need for a clock everything off a single message? and I want  of these logs generated by Systems, so I'd rather clock it off context. what do you think? If we really want to go where you are saying, we should really be clocking context's time, not here. right? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7178)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
*[drake/lcm/drake_lcm_log.h, line 38 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-KvbtfgmxYEbbjQ_ZaRC:-KvbvRXNe7pkaLSIr8sy:b48mwr4) ([raw file](https://github.com/robotlocomotion/drake/blob/215bd2c3052209bec95c49a292eafbb597f32d5a/drake/lcm/drake_lcm_log.h#L38)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Is there a way to add a connection to publish to another `DrakeLcmInterface` (namely `DrakeLcm`) such that it is easier to inspect the behavior of the playback?
</blockquote></details>

I think that should happen on the LcmPublisherSystem, namely you pass multiple interfaces to that, and leave this simple. what do you think? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7178)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
*[drake/lcm/drake_lcm_log.h, line 38 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-KvbtfgmxYEbbjQ_ZaRC:-KvbvarYkI59y8MPF32q:balql0y) ([raw file](https://github.com/robotlocomotion/drake/blob/215bd2c3052209bec95c49a292eafbb597f32d5a/drake/lcm/drake_lcm_log.h#L38)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

I think that should happen on the LcmPublisherSystem, namely you pass multiple interfaces to that, and leave this simple. what do you think? 
</blockquote></details>

I know you want to chain it here, but I think that could be confusing. If you pass multiple interfaces to publisher's ctor, that would be super obvious, but you do need to change the diagram code. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7178)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 0 of 14 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 169 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-KvcC8Ovq-eKKdnIW6JK:-KvcC8Ow-i7854uG3C0S:b-2t8q0b) ([raw file](https://github.com/robotlocomotion/drake/blob/215bd2c3052209bec95c49a292eafbb597f32d5a/drake/systems/lcm/lcm_subscriber_system.cc#L169)):*
> ```cpp
>     }
>   } else {
>     *time = lcm_interface_->GetNextMessageTime();
> ```

BTW To clarify, does this mean that every subscriber to the system will add a new event based on _any_ next message in the log?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7178)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
*[drake/systems/lcm/lcm_subscriber_system.cc, line 169 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-KvcC8Ovq-eKKdnIW6JK:-KvcF2edJRuvQBiscERl:b-h5cr7q) ([raw file](https://github.com/robotlocomotion/drake/blob/215bd2c3052209bec95c49a292eafbb597f32d5a/drake/systems/lcm/lcm_subscriber_system.cc#L169)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW To clarify, does this mean that every subscriber to the system will add a new event based on _any_ next message in the log?
</blockquote></details>

correct, whoever gets to it first will do the dispatch and advance. the rest will become a noop. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7178)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Reviewed 13 of 14 files at r1.
Review status: 13 of 14 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/lcm/drake_lcm_log.h, line 18 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kvh25PcO7PxHUXAiR6P:-Kvh25PcO7PxHUXAiR6Q:b3kre) ([raw file](https://github.com/robotlocomotion/drake/blob/215bd2c3052209bec95c49a292eafbb597f32d5a/drake/lcm/drake_lcm_log.h#L18)):*
> ```objc
> namespace lcm {
> 
> // TODO(siyuan):
> ```

?

---

*[drake/lcm/drake_lcm_log.h, line 21 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kvh2AQYRuz31NU3_byB:-Kvh2AQYRuz31NU3_byC:b-1smc0d) ([raw file](https://github.com/robotlocomotion/drake/blob/215bd2c3052209bec95c49a292eafbb597f32d5a/drake/lcm/drake_lcm_log.h#L21)):*
> ```objc
> 
> /**
>  * A Lcm interface for logging Lcm messages to a file or playing back from a
> ```

BTW: nit - I would stick to LCM or lcm. Not Lcm. 

---

*[drake/lcm/drake_lcm_log.h, line 22 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kvh2XDEDbnBT4OYx4Hi:-Kvh2XDEDbnBT4OYx4Hj:bn41fxa) ([raw file](https://github.com/robotlocomotion/drake/blob/215bd2c3052209bec95c49a292eafbb597f32d5a/drake/lcm/drake_lcm_log.h#L22)):*
> ```objc
> /**
>  * A Lcm interface for logging Lcm messages to a file or playing back from a
>  * existing log. Note the user is responsible for synchronizing the clock used
> ```

The note on synchronizing the clock needs better or more detailed explanation even at this point. 

---

*[drake/lcm/drake_lcm_log.h, line 23 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kvh2kSeyGvTRC7CisDq:-Kvh2kSeyGvTRC7CisDr:b1mccdi) ([raw file](https://github.com/robotlocomotion/drake/blob/215bd2c3052209bec95c49a292eafbb597f32d5a/drake/lcm/drake_lcm_log.h#L23)):*
> ```objc
>  * A Lcm interface for logging Lcm messages to a file or playing back from a
>  * existing log. Note the user is responsible for synchronizing the clock used
>  * to generate the log and the clock used for playback.
> ```

I would strongly recommend a more detailed doxy comment block for this class in general. This will get a lot of scrutiny from outside users..

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7178)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
*[drake/lcm/drake_lcm_log.h, line 23 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kvh2kSeyGvTRC7CisDq:-Kvh5mKvdM7fmKWIBEYs:b-4yvgou) ([raw file](https://github.com/robotlocomotion/drake/blob/215bd2c3052209bec95c49a292eafbb597f32d5a/drake/lcm/drake_lcm_log.h#L23)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

I would strongly recommend a more detailed doxy comment block for this class in general. This will get a lot of scrutiny from outside users..
</blockquote></details>

what do you want to see here? for example? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7178)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Review status: 13 of 14 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/lcm/drake_lcm_log.h, line 23 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kvh2kSeyGvTRC7CisDq:-KvhAt0zaG9X6RM6OG6E:b-5cbl3c) ([raw file](https://github.com/robotlocomotion/drake/blob/215bd2c3052209bec95c49a292eafbb597f32d5a/drake/lcm/drake_lcm_log.h#L23)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

what do you want to see here? for example? 
</blockquote></details>

Decided that it wont be necessary for now from f2f.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7178)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 13 of 14 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/lcm/drake_lcm_interface.h, line 56 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kvbty_sRPeSguEhdpS6:-KvhOHoTQSBvz2zy24au:b-3sayra) ([raw file](https://github.com/robotlocomotion/drake/blob/215bd2c3052209bec95c49a292eafbb597f32d5a/drake/lcm/drake_lcm_interface.h#L56)):*
> at first, also I don't think the system framework is designed for distributed whatever

From my understanding, yeah, the system framework is not designed for distributed-ness, but since we're using LCM (and the focus of this feature is logging LCM behavior, rather than recording the simulation in it's entirety - and IMO, should stay explicitly within this focus), it seems like it's meant for interacting with distributed systems, and possibly reproducing / catching errors that could either come from the distributed system side or internal to the system.

> If we really want to go where you are saying, we should really be clocking context's time, not here. right?

Sorry if you had mentioned this in the f2f yesterday, but I'm not sure what you mean by "clocking context's time" - do you mean having the entire sim driven by the log playback, like `/use_sim_time` in ROS? If so, that's not my intent in this comment - but rather focusing on preserving system timestamps to ensure that recording is similar to what would possibly happen on hardware, with or without Drake recording.


Overall, though, I'd be satisfied with this PR as a first pass if a TODO were added to (a) test this workflow with recording in simulation vs hardware, and (b) consider rosbag playback as a workflow that we may want to support (and possibly prototype in LCM) - would that be OK with you?

---

*[drake/lcm/drake_lcm_log.h, line 38 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-KvbtfgmxYEbbjQ_ZaRC:-KvhNTc9gA_sPljVreOF:b-io79jz) ([raw file](https://github.com/robotlocomotion/drake/blob/215bd2c3052209bec95c49a292eafbb597f32d5a/drake/lcm/drake_lcm_log.h#L38)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

I know you want to chain it here, but I think that could be confusing. If you pass multiple interfaces to publisher's ctor, that would be super obvious, but you do need to change the diagram code. 
</blockquote></details>

BTW I disagree (the obviousness comes in through how the LCM log playback is constructed) and makes it more user-friendly. However, this could be something that can be added in later (either on the Publisher side or the LcmInterface side) if need be.

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 169 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-KvcC8Ovq-eKKdnIW6JK:-KvhNoGB4IFCjGXr-NXD:b-d39p2a) ([raw file](https://github.com/robotlocomotion/drake/blob/215bd2c3052209bec95c49a292eafbb597f32d5a/drake/systems/lcm/lcm_subscriber_system.cc#L169)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

correct, whoever gets to it first will do the dispatch and advance. the rest will become a noop. 
</blockquote></details>

Is that the case?
Since `GetNextMessageTime()` does not advance time, it seems like each and every subscriber would queue up  the same publish event, but then the execution would be roughly the same once the events are executed.

Could you add a comment clarifying this?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7178)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Reviewed 14 of 14 files at r1.
Review status: 12 of 15 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/lcm/drake_lcm_interface.h, line 56 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kvbty_sRPeSguEhdpS6:-Kvm-zHgoYqD_VPelxSE:b-ex5z2l) ([raw file](https://github.com/robotlocomotion/drake/blob/215bd2c3052209bec95c49a292eafbb597f32d5a/drake/lcm/drake_lcm_interface.h#L56)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

> at first, also I don't think the system framework is designed for distributed whatever

From my understanding, yeah, the system framework is not designed for distributed-ness, but since we're using LCM (and the focus of this feature is logging LCM behavior, rather than recording the simulation in it's entirety - and IMO, should stay explicitly within this focus), it seems like it's meant for interacting with distributed systems, and possibly reproducing / catching errors that could either come from the distributed system side or internal to the system.

> If we really want to go where you are saying, we should really be clocking context's time, not here. right?

Sorry if you had mentioned this in the f2f yesterday, but I'm not sure what you mean by "clocking context's time" - do you mean having the entire sim driven by the log playback, like `/use_sim_time` in ROS? If so, that's not my intent in this comment - but rather focusing on preserving system timestamps to ensure that recording is similar to what would possibly happen on hardware, with or without Drake recording.


Overall, though, I'd be satisfied with this PR as a first pass if a TODO were added to (a) test this workflow with recording in simulation vs hardware, and (b) consider rosbag playback as a workflow that we may want to support (and possibly prototype in LCM) - would that be OK with you?
</blockquote></details>

as a compromise, i added a flag in the ctor to overwrite log timestamp with system clock. 

---

*[drake/lcm/drake_lcm_log.h, line 18 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kvh25PcO7PxHUXAiR6P:-KvlvksZo63qR-rIjFAP:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/215bd2c3052209bec95c49a292eafbb597f32d5a/drake/lcm/drake_lcm_log.h#L18)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

?
</blockquote></details>

Done.

---

*[drake/lcm/drake_lcm_log.h, line 21 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kvh2AQYRuz31NU3_byB:-KvlvpQWmy3U0btw0o9R:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/215bd2c3052209bec95c49a292eafbb597f32d5a/drake/lcm/drake_lcm_log.h#L21)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

BTW: nit - I would stick to LCM or lcm. Not Lcm. 
</blockquote></details>

Done.

---

*[drake/lcm/drake_lcm_log.h, line 22 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kvh2XDEDbnBT4OYx4Hi:-Kvlvpu39fXcFlz1dY8q:b-t9490b) ([raw file](https://github.com/robotlocomotion/drake/blob/215bd2c3052209bec95c49a292eafbb597f32d5a/drake/lcm/drake_lcm_log.h#L22)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

The note on synchronizing the clock needs better or more detailed explanation even at this point. 
</blockquote></details>

added 

---

*[drake/lcm/drake_lcm_log.h, line 38 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-KvbtfgmxYEbbjQ_ZaRC:-Kvlw4pXq8aJuJEcFubW:b-ykd769) ([raw file](https://github.com/robotlocomotion/drake/blob/215bd2c3052209bec95c49a292eafbb597f32d5a/drake/lcm/drake_lcm_log.h#L38)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW I disagree (the obviousness comes in through how the LCM log playback is constructed) and makes it more user-friendly. However, this could be something that can be added in later (either on the Publisher side or the LcmInterface side) if need be.
</blockquote></details>

I still think chaining multiple interfaces together is an interesting idea, but I want to enable that work flow in Publisher's ctor, not in the interfaces. It's obvious to do Publish on a chained interface, I just publish to all of them, it's not obvious what do to with subscribe or callbacks. Chaining interfaces together makes that confusing. If you want multiple publish, should pass multiple interfaces to publisher. added a todo in lcm subscriber. 

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 169 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-KvcC8Ovq-eKKdnIW6JK:-KvhmMQuuWvRpvbQSybs:b-v4mw52) ([raw file](https://github.com/robotlocomotion/drake/blob/215bd2c3052209bec95c49a292eafbb597f32d5a/drake/systems/lcm/lcm_subscriber_system.cc#L169)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Is that the case?
Since `GetNextMessageTime()` does not advance time, it seems like each and every subscriber would queue up  the same publish event, but then the execution would be roughly the same once the events are executed.

Could you add a comment clarifying this?
</blockquote></details>

you are correct about every subscriber will scheduled an publish event for themselves. however when they start executing the callback, see below, they would be looping over lcm_interface_->GetNextMessageTime() == c.get_time(), and only do stuff if the context's time at callback is the same as the log's front's time. once any subscriber advanced the log till a time that doesn't 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7178)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 12 of 15 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-:-Kvmch4mdeuy4hTKc8PC:bnfp4nl)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 12 of 15 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 169 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-KvcC8Ovq-eKKdnIW6JK:-Kvmoiyb-9q-3ygwgyF5:b9duepp) ([raw file](https://github.com/robotlocomotion/drake/blob/215bd2c3052209bec95c49a292eafbb597f32d5a/drake/systems/lcm/lcm_subscriber_system.cc#L169)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

you are correct about every subscriber will scheduled an publish event for themselves. however when they start executing the callback, see below, they would be looping over lcm_interface_->GetNextMessageTime() == c.get_time(), and only do stuff if the context's time at callback is the same as the log's front's time. once any subscriber advanced the log till a time that doesn't 
</blockquote></details>

I see, thanks for the explanation! Could you capture in a quick comment here, if possible?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7178)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Reviewed 3 of 3 files at r2.
Review status: 14 of 15 files reviewed at latest revision, all discussions resolved.

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 169 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-KvcC8Ovq-eKKdnIW6JK:-Kvmwv7uFxRxcjMIly_x:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/215bd2c3052209bec95c49a292eafbb597f32d5a/drake/systems/lcm/lcm_subscriber_system.cc#L169)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

I see, thanks for the explanation! Could you capture in a quick comment here, if possible?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7178)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7178)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> pending issues noted below.

---

Reviewed 5 of 14 files at r1, 2 of 3 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/lcm/drake_lcm_log.h, line 23 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kvn7Udu6-wFQhVuTNKC:-Kvn7Udu6-wFQhVuTNKD:b-8w21t5) ([raw file](https://github.com/robotlocomotion/drake/blob/bf24558a99ed7b01475c20d33dea90bed54e1f26/drake/lcm/drake_lcm_log.h#L23)):*
> ```objc
>  * to generate the log and the clock used for playback. For example, if the log
>  * is generated by some external logger (the lcm-logger binary), which uses the
>  * unix epoc time clock to record message arrival time, the user need to offset
> ```

FYI -- Typo "epoc" should be "epoch"

---

*[drake/lcm/test/drake_lcm_log_test.cc, line 54 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kvn7z3DaFOwMTVsvZhE:-Kvn7z3DaFOwMTVsvZhF:bwy4515) ([raw file](https://github.com/robotlocomotion/drake/blob/bf24558a99ed7b01475c20d33dea90bed54e1f26/drake/lcm/test/drake_lcm_log_test.cc#L54)):*
> ```cpp
>   // Finish writing.
>   w_log.reset();
> ```

FYI extra newline.

---

*[drake/systems/lcm/test/lcm_log_playback_test.cc, line 18 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kvn8F6dWm6Wi_L_xFN8:-Kvn8F6eqrWLA9r2W6iW:byxc27m) ([raw file](https://github.com/robotlocomotion/drake/blob/bf24558a99ed7b01475c20d33dea90bed54e1f26/drake/systems/lcm/test/lcm_log_playback_test.cc#L18)):*
> ```cpp
> 
> void SetInput(double time, const std::string& name, double val,
>     Context<double>* context) {
> ```

FYI this indentation looks wrong to me; check it.

---

*[drake/systems/lcm/test/lcm_log_playback_test.cc, line 123 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kvn8bw2SDHdZT0dYA8Z:-Kvn8bw2SDHdZT0dYA8_:b-d3vo0a) ([raw file](https://github.com/robotlocomotion/drake/blob/bf24558a99ed7b01475c20d33dea90bed54e1f26/drake/systems/lcm/test/lcm_log_playback_test.cc#L123)):*
> ```cpp
>     // msg.timestamp is generated with the shifted time.
>     EXPECT_NEAR(msg.timestamp, expected_times[i] * 1e6, 1e-12);
>     // Note: this needs to match the how we schedule event processing
> ```

FYI typo "the how"

---

*[drake/systems/lcm/test/lcm_log_playback_test.cc, line 125 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kvn8vYvVBy2MGT_pqD5:-Kvn8vYvVBy2MGT_pqD6:b96uk8k) ([raw file](https://github.com/robotlocomotion/drake/blob/bf24558a99ed7b01475c20d33dea90bed54e1f26/drake/systems/lcm/test/lcm_log_playback_test.cc#L125)):*
> ```cpp
>     // Note: this needs to match the how we schedule event processing
>     // (DoCalcNextUpdateTime) in LcmSubscriberSystem.
>     EXPECT_NEAR(times[i], expected_times[i] + 0.0001, 1e-12);
> ```

Since this magic number needs to exactly match a similar magic number in `lcm_subscriber_system.cc`, it would be better to move that magic number into the header as a public constant.  Alternatively mark this as TODO and reference #5725 so at least there is a chance that whoever fixes 5725 sees it.

---

*[drake/systems/lcm/test/lcm_log_playback_test.cc, line 182 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kvn9UmW1unbc5nqnXIT:-Kvn9UmW1unbc5nqnXIU:b-qgbsyj) ([raw file](https://github.com/robotlocomotion/drake/blob/bf24558a99ed7b01475c20d33dea90bed54e1f26/drake/systems/lcm/test/lcm_log_playback_test.cc#L182)):*
> ```cpp
> // subscriber, the content of each distinct message and its receive time are
> // recorded and compared against expected values.
> GTEST_TEST(haha, testhaha) {
> ```

BTW This doesn't really seem like an informative name for this test case.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-:-Kvn9cMp81UIdBO6Dpxn:b-mmtvc7)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
Review status: 12 of 15 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/lcm/drake_lcm_log.h, line 38 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-KvbtfgmxYEbbjQ_ZaRC:-Kvn5DdyzyjHdjM4TihE:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/215bd2c3052209bec95c49a292eafbb597f32d5a/drake/lcm/drake_lcm_log.h#L38)):*
<details><summary><i>Previously, siyuanfeng-tri wrote…</i></summary><blockquote>

I still think chaining multiple interfaces together is an interesting idea, but I want to enable that work flow in Publisher's ctor, not in the interfaces. It's obvious to do Publish on a chained interface, I just publish to all of them, it's not obvious what do to with subscribe or callbacks. Chaining interfaces together makes that confusing. If you want multiple publish, should pass multiple interfaces to publisher. added a todo in lcm subscriber. 
</blockquote></details>

Done.

---

*[drake/lcm/drake_lcm_log.h, line 23 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kvn7Udu6-wFQhVuTNKC:-KvnFOjIMqTgyjp_Brae:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/bf24558a99ed7b01475c20d33dea90bed54e1f26/drake/lcm/drake_lcm_log.h#L23)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

FYI -- Typo "epoc" should be "epoch"
</blockquote></details>

Done.

---

*[drake/lcm/test/drake_lcm_log_test.cc, line 54 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kvn7z3DaFOwMTVsvZhE:-KvnFUQhlhfoq2w0G25l:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/bf24558a99ed7b01475c20d33dea90bed54e1f26/drake/lcm/test/drake_lcm_log_test.cc#L54)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

FYI extra newline.
</blockquote></details>

Done.

---

*[drake/systems/lcm/test/lcm_log_playback_test.cc, line 18 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kvn8F6dWm6Wi_L_xFN8:-KvnFhXnoxVW3ymWPDO7:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/bf24558a99ed7b01475c20d33dea90bed54e1f26/drake/systems/lcm/test/lcm_log_playback_test.cc#L18)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

FYI this indentation looks wrong to me; check it.
</blockquote></details>

Done.

---

*[drake/systems/lcm/test/lcm_log_playback_test.cc, line 123 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kvn8bw2SDHdZT0dYA8Z:-KvnFm8r2ugPnA0wrnnD:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/bf24558a99ed7b01475c20d33dea90bed54e1f26/drake/systems/lcm/test/lcm_log_playback_test.cc#L123)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

FYI typo "the how"
</blockquote></details>

Done.

---

*[drake/systems/lcm/test/lcm_log_playback_test.cc, line 125 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kvn8vYvVBy2MGT_pqD5:-KvnG4TFBQiyH5rgKB-C:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/bf24558a99ed7b01475c20d33dea90bed54e1f26/drake/systems/lcm/test/lcm_log_playback_test.cc#L125)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

Since this magic number needs to exactly match a similar magic number in `lcm_subscriber_system.cc`, it would be better to move that magic number into the header as a public constant.  Alternatively mark this as TODO and reference #5725 so at least there is a chance that whoever fixes 5725 sees it.
</blockquote></details>

Done.

---

*[drake/systems/lcm/test/lcm_log_playback_test.cc, line 182 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kvn9UmW1unbc5nqnXIT:-KvnGVhsSNzBGPZ8ou_c:b-rs0lao) ([raw file](https://github.com/robotlocomotion/drake/blob/bf24558a99ed7b01475c20d33dea90bed54e1f26/drake/systems/lcm/test/lcm_log_playback_test.cc#L182)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

BTW This doesn't really seem like an informative name for this test case.
</blockquote></details>

fail. forgot to change name. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7178)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 5 of 14 files at r1, 1 of 3 files at r2, 1 of 1 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7178)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
#7217 was assigned to me, so I went back to read the original PR that its fixing.

@siyuanfeng-tri Please take a look below, and resolve all of these defects in a follow-up PR.

---

Reviewed 9 of 14 files at r1, 2 of 3 files at r2, 1 of 1 files at r3, 2 of 3 files at r4.
Review status: all files reviewed at latest revision, 26 unresolved discussions, all commit checks successful.

---

*[drake/lcm/drake_lcm.h, line 43 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kw0Ijtr2taRgVTWByvA:-Kw0IjtsFiT7Wttk30V7:byvsbrd) ([raw file](https://github.com/robotlocomotion/drake/blob/01c98a864d7fb92eadcdc8263d9b28278da4928f/drake/lcm/drake_lcm.h#L43)):*
> ```objc
> 
>   void Publish(const std::string& channel, const void* data,
>                int data_size, double time_sec = 0) override;
> ```

Brittle repetition of default value.

---

*[drake/lcm/drake_lcm_interface.h, line 53 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kw0ILIbLIkPTYJV32es:-Kw0ILIbLIkPTYJV32et:ba4wfue) ([raw file](https://github.com/robotlocomotion/drake/blob/01c98a864d7fb92eadcdc8263d9b28278da4928f/drake/lcm/drake_lcm_interface.h#L53)):*
> ```objc
>    *
>    * @param[in] time_sec Time in seconds when the publish event occurred. Note
>    * that this argument is only used when generating a Lcm log.
> ```

The second sentence is overly-constraining, and thus brittle.  In fact, the default publisher always sets this value to non-zero, and nothing would stop other implementations of this interface from using that value.

---

*[drake/lcm/drake_lcm_interface.h, line 56 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kw0IRCd5v7Q_DJGdtyt:-Kw0IRCd5v7Q_DJGdtyu:b-4xy1xc) ([raw file](https://github.com/robotlocomotion/drake/blob/01c98a864d7fb92eadcdc8263d9b28278da4928f/drake/lcm/drake_lcm_interface.h#L56)):*
> ```objc
>    */
>   virtual void Publish(const std::string& channel, const void* data,
>                        int data_size, double time_sec = 0) = 0;
> ```

Never use default values in virtual methods.  They don't survive overrides robustly.

---

*[drake/lcm/drake_lcm_interface.h, line 71 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kw0I6jb9gGr3wpObIjt:-Kw0I6jc92NI5ENdOfmH:blnbg2b) ([raw file](https://github.com/robotlocomotion/drake/blob/01c98a864d7fb92eadcdc8263d9b28278da4928f/drake/lcm/drake_lcm_interface.h#L71)):*
> ```objc
> 
>   /**
>    * Only used for supporting Lcm log playback. Default implementation is no-op.
> ```

This fails to document what the method actually means or does.

---

*[drake/lcm/drake_lcm_interface.h, line 71 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kw0Iqi_USMEFCeKQZmS:-Kw0Iqi_USMEFCeKQZmT:b5tz5up) ([raw file](https://github.com/robotlocomotion/drake/blob/01c98a864d7fb92eadcdc8263d9b28278da4928f/drake/lcm/drake_lcm_interface.h#L71)):*
> ```objc
> 
>   /**
>    * Only used for supporting Lcm log playback. Default implementation is no-op.
> ```

The first sentence is overly constraining.

---

*[drake/lcm/drake_lcm_interface.h, line 71 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kw0IuiQNcgoNOgR0f4_:-Kw0IuiRrkLWmrIjVod8:b-azlt9r) ([raw file](https://github.com/robotlocomotion/drake/blob/01c98a864d7fb92eadcdc8263d9b28278da4928f/drake/lcm/drake_lcm_interface.h#L71)):*
> ```objc
> 
>   /**
>    * Only used for supporting Lcm log playback. Default implementation is no-op.
> ```

_META_ Don't say `Lcm`, say `LCM` for consistency within the file.

---

*[drake/lcm/drake_lcm_interface.h, line 74 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kw0IyUFVWWHYgjxkyzp:-Kw0IyUFVWWHYgjxkyzq:b-97iloq) ([raw file](https://github.com/robotlocomotion/drake/blob/01c98a864d7fb92eadcdc8263d9b28278da4928f/drake/lcm/drake_lcm_interface.h#L74)):*
> ```objc
>    * @see DrakeLcmLog::DispatchMessageAndAdvanceLog
>    */
>   virtual void DispatchMessageAndAdvanceLog(double) {}
> ```

Unclear what the argument value means; it is unnamed and undocumented.

---

*[drake/lcm/drake_lcm_interface.h, line 77 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kw0I3CTMoPZQ2_J8JvM:-Kw0I3CUfKxueDBNiD8u:blnbg2b) ([raw file](https://github.com/robotlocomotion/drake/blob/01c98a864d7fb92eadcdc8263d9b28278da4928f/drake/lcm/drake_lcm_interface.h#L77)):*
> ```objc
> 
>   /**
>    * Only used for supporting Lcm log playback. Default implementation returns
> ```

This fails to document what the method actually means or does.

---

*[drake/lcm/drake_lcm_interface.h, line 77 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kw0IsX-aAv86v_iKq8U:-Kw0IsX-aAv86v_iKq8V:b5tz5up) ([raw file](https://github.com/robotlocomotion/drake/blob/01c98a864d7fb92eadcdc8263d9b28278da4928f/drake/lcm/drake_lcm_interface.h#L77)):*
> ```objc
> 
>   /**
>    * Only used for supporting Lcm log playback. Default implementation returns
> ```

The first sentence is overly constraining.

---

*[drake/lcm/drake_lcm_log.cc, line 23 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kw0OEuRFmijvf4DZ_cQ:-Kw0OEuSQXgkHSPKD-N_:bagret) ([raw file](https://github.com/robotlocomotion/drake/blob/01c98a864d7fb92eadcdc8263d9b28278da4928f/drake/lcm/drake_lcm_log.cc#L23)):*
> ```cpp
>   }
>   if (!log_->good()) {
>     throw std::runtime_error("Failed to open log file: " + file_name);
> ```

FYI Consider adding a unit test for this.  I'm not sure what `readNextEvent` does when the file was not found.

---

*[drake/lcm/drake_lcm_log.cc, line 33 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kw0ORa-eNgoFZIQrAOC:-Kw0ORa10gno-twX8zGe:b-yl8kar) ([raw file](https://github.com/robotlocomotion/drake/blob/01c98a864d7fb92eadcdc8263d9b28278da4928f/drake/lcm/drake_lcm_log.cc#L33)):*
> ```cpp
>   }
> 
>   std::lock_guard<std::mutex> lock(mutex_);
> ```

It is very much not clear to my where multiple threads are coming from in terms of this class.  That needs to be documented somewhere.

---

*[drake/lcm/drake_lcm_log.cc, line 35 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kw0OqoI2fTCixY3FWH0:-Kw0OqoJyVyJ_z6aV6az:byni03o) ([raw file](https://github.com/robotlocomotion/drake/blob/01c98a864d7fb92eadcdc8263d9b28278da4928f/drake/lcm/drake_lcm_log.cc#L35)):*
> ```cpp
>   std::lock_guard<std::mutex> lock(mutex_);
> 
>   ::lcm::LogEvent log_event;
> ```

FYI LCM Doesn't value-initialize its structs; suggest `log_event{};` here so that we know the memory is not garbage.

---

*[drake/lcm/drake_lcm_log.cc, line 48 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kw0P7eE0cAcb3cg7N2q:-Kw0P7eE0cAcb3cg7N2r:byb76da) ([raw file](https://github.com/robotlocomotion/drake/blob/01c98a864d7fb92eadcdc8263d9b28278da4928f/drake/lcm/drake_lcm_log.cc#L48)):*
> ```cpp
>   log_event.data = const_cast<void*>(data);
> 
>   // TODO(siyuan): should make cache this or thread write this.
> ```

BTW bad grammar `make cache this`

---

*[drake/lcm/drake_lcm_log.h, line 23 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kw0N5mvtvUk0MUJZkOS:-Kw0N5mvtvUk0MUJZkOT:bmlfark) ([raw file](https://github.com/robotlocomotion/drake/blob/01c98a864d7fb92eadcdc8263d9b28278da4928f/drake/lcm/drake_lcm_log.h#L23)):*
> ```objc
>  * to generate the log and the clock used for playback. For example, if the log
>  * is generated by some external logger (the lcm-logger binary), which uses the
>  * unix epoch time clock to record message arrival time, the user need to offset
> ```

BTW nit Typo `needs`.

---

*[drake/lcm/drake_lcm_log.h, line 45 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kw0MxdFWg8Wb2Ta-f2a:-Kw0MxdFWg8Wb2Ta-f2b:bjkep6s) ([raw file](https://github.com/robotlocomotion/drake/blob/01c98a864d7fb92eadcdc8263d9b28278da4928f/drake/lcm/drake_lcm_log.h#L45)):*
> ```objc
>    * @throws std::runtime_error if unable to open file.
>    */
>   explicit DrakeLcmLog(const std::string& file_name, bool is_write,
> ```

BTW nit Pointless `explicit`.

---

*[drake/lcm/drake_lcm_log.h, line 45 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kw0NROenkZFpt89NQ5V:-Kw0NROfBZ2rDLcfPfE-:bolosb1) ([raw file](https://github.com/robotlocomotion/drake/blob/01c98a864d7fb92eadcdc8263d9b28278da4928f/drake/lcm/drake_lcm_log.h#L45)):*
> ```objc
>    * @throws std::runtime_error if unable to open file.
>    */
>   explicit DrakeLcmLog(const std::string& file_name, bool is_write,
> ```

FYI _Meta_ Consider `is_write_only` to be consistent with the rest of the class API.

---

*[drake/lcm/drake_lcm_log.h, line 129 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kw0MrsqNCgiphW14IPw:-Kw0MrsrnlUQQegF0yuM:b-r6hj1b) ([raw file](https://github.com/robotlocomotion/drake/blob/01c98a864d7fb92eadcdc8263d9b28278da4928f/drake/lcm/drake_lcm_log.h#L129)):*
> ```objc
>       subscriptions_;
> 
>   mutable std::mutex mutex_;
> ```

Missing documentation for which member data it is guarding.

---

*[drake/lcm/drake_mock_lcm.h, line 44 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kw0J3pUKKK9d8L0kNBH:-Kw0J3pUKKK9d8L0kNBI:b8ub9ts) ([raw file](https://github.com/robotlocomotion/drake/blob/01c98a864d7fb92eadcdc8263d9b28278da4928f/drake/lcm/drake_mock_lcm.h#L44)):*
> ```objc
> 
>   void Publish(const std::string& channel, const void* data,
>                int data_size, double time_sec = 0) override;
> ```

Brittle duplication of default value.

---

*[drake/lcm/test/drake_lcm_log_test.cc, line 31 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kw0PXOtrFkACmNzN2ah:-Kw0PXOtrFkACmNzN2ai:b-f79z3j) ([raw file](https://github.com/robotlocomotion/drake/blob/01c98a864d7fb92eadcdc8263d9b28278da4928f/drake/lcm/test/drake_lcm_log_test.cc#L31)):*
> ```cpp
>  private:
>   const std::string name_;
>   MsgType msg_;
> ```

Missing `{}` on LCM-generated struct (it defaults to garbage).

---

*[drake/systems/lcm/BUILD, line 51 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kw0MFLpNaC5SUfHF4Wh:-Kw0MFLqWKytNr15yxir:b-lvrtp0) ([raw file](https://github.com/robotlocomotion/drake/blob/01c98a864d7fb92eadcdc8263d9b28278da4928f/drake/systems/lcm/BUILD#L51)):*
> ```Python
>         ":translator",
>         "//drake/lcm:interface",
>         "//drake/lcm:lcm_log",
> ```

This seems spurious.

---

*[drake/systems/lcm/lcm_publisher_system.cc, line 106 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kw0JTnfcVXgyYDhvo9b:-Kw0JTngwtHva-1PinhJ:b-cupdn1) ([raw file](https://github.com/robotlocomotion/drake/blob/01c98a864d7fb92eadcdc8263d9b28278da4928f/drake/systems/lcm/lcm_publisher_system.cc#L106)):*
> ```cpp
>   // Publishes onto the specified LCM channel.
>   lcm_->Publish(channel_, message_bytes.data(), message_bytes.size(),
>                 context.get_time());
> ```

Missing direct unit test coverage that the time being passed here is correct!

---

*[drake/systems/lcm/lcm_publisher_system.h, line 46 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kw0JDFG5GMsYTdg9eLn:-Kw0JDFG5GMsYTdg9eLo:b5tet52) ([raw file](https://github.com/robotlocomotion/drake/blob/01c98a864d7fb92eadcdc8263d9b28278da4928f/drake/systems/lcm/lcm_publisher_system.h#L46)):*
> ```objc
> 
>   // TODO(siyuan): add multiple DrakeLcmInterface, so you can publish to a log
>   // and real lcm at the same time.
> ```

FYI This isn't a great place for this TODO.  The solution will not touch this file, so its likely that the author would forget to update this site.  Why not use a Drake issue instead (or maybe just not write it down; someone will do the work exactly when they need it).  TODOs are best for capturing caveats with a current implementation, not writing up a project plan.

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 174 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kw0KEOMjYu1TIJxJCwf:-Kw0KEOMjYu1TIJxJCwg:b-k41ip4) ([raw file](https://github.com/robotlocomotion/drake/blob/01c98a864d7fb92eadcdc8263d9b28278da4928f/drake/systems/lcm/lcm_subscriber_system.cc#L174)):*
> ```cpp
>     DRAKE_DEMAND(*time > context.get_time());
>     if (std::isinf(*time)) {
>       return;
> ```

Why did we stop calling the base class' implementation?  Our class is not final, so we cannot assume that there are no other, e.g., periodic updates that need to be processed.

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 194 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kw0L2E31MLBHbs2ucSm:-Kw0L2E4-pq8ms6Dy0kg:bbwoxei) ([raw file](https://github.com/robotlocomotion/drake/blob/01c98a864d7fb92eadcdc8263d9b28278da4928f/drake/systems/lcm/lcm_subscriber_system.cc#L194)):*
> ```cpp
>         // Want to keep polling from the message queue, if they happen
>         // to be occur at the exact same time.
>         while (lcm_interface_->GetNextMessageTime() == c.get_time()) {
> ```

BTW It seems like some kind of invariant is happening here, but isn't documented with an assertion.  Something like `c.get_time()` matches the `*time` that we asked for, and that `GetNextMessageTime()` is either `c.get_time()` or later?

---

*[drake/systems/lcm/lcm_subscriber_system.cc, line 203 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kw0LNFoGidbE5KSMFUF:-Kw0LNFoGidbE5KSMFUG:b-1n4ii1) ([raw file](https://github.com/robotlocomotion/drake/blob/01c98a864d7fb92eadcdc8263d9b28278da4928f/drake/systems/lcm/lcm_subscriber_system.cc#L203)):*
> ```cpp
>           Event<double>::TriggerType::kTimed,
>           callback));
>   }
> ```

Even though there is a nice round-trip acceptance test that uses DrakeLcmLog, it seems to me like the new code here should have direct unit test coverage in lcm_subscriber_system_test.cc.

Or at a minimum, should have a comment in that file (or here) explanining where the test coverage comes from, for future maintainers.

---

*[drake/systems/lcm/lcm_subscriber_system.h, line 220 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-Kw0JsQbFD7BVXv7W9Wd:-Kw0JsQbFD7BVXv7W9We:brq3vyb) ([raw file](https://github.com/robotlocomotion/drake/blob/01c98a864d7fb92eadcdc8263d9b28278da4928f/drake/systems/lcm/lcm_subscriber_system.h#L220)):*
> ```objc
>   int received_message_count_{0};
> 
>   drake::lcm::DrakeLcmInterface* lcm_interface_;
> ```

Missing value-initialization of non-const POD member field.

(Probably the right answer is to mark it const (after the `*`), and then for clarity move it higher in the list of member fields, since its a constructor-time initialize-once invariant, not ever-changing member data.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7178#-:-Kw0JgPcJKLq3vcbLnw9:b-xg4obo)*
<!-- Sent from Reviewable.io -->
